modules/axnet/src/smoltcp_impl/mod.rs

```diff
@@ -180,45 +169,41 @@ impl DeviceWrapper {
     fn new(inner: AxNetDevice) -> Self {
         Self {
             inner: RefCell::new(inner),
-            rx_buf_queue: VecDeque::with_capacity(RX_BUF_QUEUE_SIZE),
-        }
-    }
-
-    fn poll<F>(&mut self, f: F)
-    where
-        F: Fn(&[u8]),
-    {
-        while self.rx_buf_queue.len() < RX_BUF_QUEUE_SIZE {
-            match self.inner.borrow_mut().receive() {
-                Ok(buf) => {
-                    f(buf.packet());
-                    self.rx_buf_queue.push_back(buf);
-                }
-                Err(DevError::Again) => break, // TODO: better method to avoid error type conversion
-                Err(err) => {
-                    warn!("receive failed: {:?}", err);
-                    break;
-                }
-            }
         }
     }
-
-    fn receive(&mut self) -> Option<NetBufferBox<'static>> {
-        self.rx_buf_queue.pop_front()
-    }
 }
 
 impl Device for DeviceWrapper {
     type RxToken<'a> = AxNetRxToken<'a> where Self: 'a;
     type TxToken<'a> = AxNetTxToken<'a> where Self: 'a;
 
     fn receive(&mut self, _timestamp: Instant) -> Option<(Self::RxToken<'_>, Self::TxToken<'_>)> {
-        let rx_buf = self.receive()?;
-        Some((AxNetRxToken(&self.inner, rx_buf), AxNetTxToken(&self.inner)))
+        let mut dev = self.inner.borrow_mut();
+        dev.recycle_tx_bufs().unwrap();
+        if dev.can_receive() && dev.can_transmit() {
```

Comment:
`dev.can_receive` is duplicated with `match dev.receive()`, simplify as follows:

```rust
if !dev.can_transmit() {
    return None;
}
let rx_buf = match dev.receive() { ... };
```